### DISCUSSION TITLE
Runtime logging with transform messages confusion

### DISCUSSION THREAD

**Discussion by Nathan-Intergral** [microsoft/autogen#2486] [General]
Hey everyone,

I'm a bit puzzled and hope someone can shed some light on this. I've set up runtime logging in my group chat, which uses TransfromMessages to limit message tokens and history. Everything seems fine in the console and debug logs—the limits are being applied as they should. But when I check the runtime logs, it looks like the token count for each message keeps increasing. Upon closer inspection, it seems like every "request" field includes the whole chat history.

So now I'm questioning the accuracy of my logs. I'm not sure if the token counts and costs I'm getting from group_chat_manager.get_total_usage() are reliable or if they even match what's shown in the chat completion logs. Does anyone else see this happening? Am I missing something?

Is it known that the runtime logger doesnt take in to account your transforms?

**Comment by ekzhu** [comment]
runtime_logger should log the exact request sent to LLM. So it looks like the transformation is not applied to some of your agents. Can you post your code snippet? I suspect the transformation is not applied to the group chat select speaker messages. 

cc @WaelKarkoub @joshkyh @gagb @cheng-tan

**Reply by Nathan-Intergral** [reply]
Ah! Potentially, ive applied it to the overseer agent which is the user proxy in my case. Heres the code:

```python
def create_worker_suite(user, context):
    workers = init_workers(user, context)
    context_handlers = setup_context_handlers()

    medium_context_workers = [worker_models.WorkerRole.overseer.value]
    large_context_workers = [worker_models.WorkerRole.organiser.value]
    for name, worker in workers.items():
        handler = context_handlers["small"]  # Default to small context
        if name in large_context_workers:
            handler = context_handlers["large"]
        elif name in medium_context_workers:
            handler = context_handlers["medium"]
        handler.add_to_agent(worker)

    return workers

def init_workers(user, context):
    workers = {}
    worker_definitions = get_worker_definitions(user, context)
    for worker, definition in worker_definitions.items():
        workers[worker] = create_assistant_worker(
            worker, definition["description"], definition["llm_config_func"]
        )
    workers[worker_models.WorkerRole.overseer.value] = create_overseer_worker(
        user, context
    )
    return workers

def setup_context_handlers():
    return {
        "small": transform_messages.TransformMessages(
            transforms=[
                transforms.MessageHistoryLimiter(max_messages=5),
                transforms.MessageTokenLimiter(
                    max_tokens=4000,
                    max_tokens_per_message=3000,
                    model=group_chat_models.LLMModel.GPT4.value,
                ),
            ]
        ),
        "medium": transform_messages.TransformMessages(
            transforms=[
                transforms.MessageHistoryLimiter(max_messages=10),
                transforms.MessageTokenLimiter(
                    max_tokens=8000,
                    max_tokens_per_message=3000,
                    model=group_chat_models.LLMModel.GPT4.value,
                ),
            ]
        ),
        "large": transform_messages.TransformMessages(
            transforms=[
                transforms.MessageHistoryLimiter(max_messages=20),
                transforms.MessageTokenLimiter(
                    max_tokens=8000,
                    max_tokens_per_message=3000,
                    model=group_chat_models.LLMModel.GPT4.value,
                ),
            ]
        ),
    }
```

Thanks for the help!

**Reply by WaelKarkoub** [reply]
There are no tangible benefits to adding the transform messages capability to user proxies. Try to do the same thing (adding `TransformMessages` capability) but to llm-based agents.

**Reply by Nathan-Intergral** [reply]
ah thank you ill change that. However the rest of the "workers" besides the overseer are AssistantAgent's. I see what you mean about it being the speaker selection prompt specifically not being transformed. Is there a way to apply the transforms to those speaker selection prompts?

**Reply by WaelKarkoub** [reply]
How are you setting up your groupchat and groupchat manager?

**Reply by Nathan-Intergral** [reply]
``` Python
    def initialise_group_chat_manager(
        self, context, workers, messages=[], max_round=1000, introductions=False
    ) -> GroupChatManager:
        group_chat = GroupChat(
            agents=list(workers.values()),
            messages=messages,
            max_round=max_round,
            speaker_selection_method="auto",
            send_introductions=introductions,
            speaker_transitions_type="allowed",
            allowed_or_disallowed_speaker_transitions=get_allowed_transitions(workers),
        )
        context.set_group_chat(group_chat)

        manager = GroupChatManager(
            groupchat=group_chat,
            llm_config=get_llm_config(),
            is_termination_msg=lambda x: "GROUPCHAT_TERMINATE" in x.get("content", ""),
        )
        return manager
```

**Reply by WaelKarkoub** [reply]
Okay I see, I think it's starting to make sense. `GroupChatManager` is essentially another LLM agent if `speaker_selection_method` in `GroupChat` is set to `auto`. What you can do is apply `TransformMessages` to the `manager`. You might have to tune `select_speaker_prompt_template` for `group_chat` as it might not be descriptive enough for your use case.

**Reply by Nathan-Intergral** [reply]
Okay great, ill give that a go. Thanks!

**Reply by ekzhu** [reply]
@Nathan-Intergral @WaelKarkoub just for awareness @marklysze  is working on a PR to improve the select speaker LLM call to make it an agent conversation, in that case, we can apply message transformation to the inner agent. https://github.com/microsoft/autogen/pull/2304

**Reply by marklysze** [reply]
Thanks for the heads up, I'm close to committing my changes, can you let me know if I need to add something?

**Reply by ekzhu** [reply]
I don't think we need to expand the scope of that PR more.

**Reply by Nathan-Intergral** [reply]
Hey, so to update, i tried applying the message transform to the group chat manager and it didnt seem to work. I hat a little dig and i can see that the `select_speaker` function in `groupchat.py` uses `conversable_agent.py`'s `generate_oai_reply`. Which unlike `generate_reply` does not call `process_all_messages_before_reply` so the hook for transform messages is never run from what i can tell. Am i understanding this correctly?

**Reply by WaelKarkoub** [reply]
@Nathan-Intergral I believe your understanding is correct, good catch! Whenever you have time, can you open an issue and tag us? I can start working on it after https://github.com/microsoft/autogen/pull/2304 gets merged. Like @ekzhu said, we can start applying message transforms to the inner agent once that PR goes in and it might fix your problems.

**Reply by Nathan-Intergral** [reply]
@WaelKarkoub I have created issue #2499

**Reply by marklysze** [reply]
Hi guys, I have a PR for updating select speaker for auto mode that uses a two-way chat and changes the select speaker code. Can you please check that before changing select_speaker.

See #2304 

Any questions just ask :)
